### DISCUSSION TITLE
Using AuthP in WebApi is not working showing No service for type 'Microsoft.AspNetCore.Identity.UserManager

### DISCUSSION THREAD

**Discussion by nasir-adappt** [JonPSmith/AuthPermissions.AspNetCore#70] [Q&A]
I am trying to implement the AuthP library in my .NET 6 WebApi project. I have used ASP.NET Core Identity for Individual Account security. I have two custom Identity classes which are:
ApplicationUser extended from IdentityUser and
ApplicationRole extended from IdentityRole

Now I have installed the AuthP library in my project and added the below code for configuration:
```
builder.Services.RegisterAuthPermissions<UserPermissions>(
    options =>
    {
        options.UseLocksToUpdateGlobalResources = false;
    })
    .UsingEfCoreSqlServer(connectionString)
    .IndividualAccountsAuthentication<ApplicationUser>()
    .AddSuperUserToIndividualAccounts<ApplicationUser>()
    .RegisterFindUserInfoService<CustomAccountUserLookup>()
    .AddRolesPermissionsIfEmpty(AppAuthSetupData.RolesDefinition)
    .AddAuthUsersIfEmpty(AppAuthSetupData.UsersRolesDefinition)
    .SetupAspNetCoreAndDatabase(options =>
    {
        //Migrate individual account database
        options.RegisterServiceToRunInJob<StartupServiceMigrateAnyDbContext<ApplicationDbContext>>();
        //Add demo users to the database
        options.RegisterServiceToRunInJob<StartupServicesIndividualAccountsAddDemoUsers>();
    });
```
But whenever I run the project I get the below error:
**No service for type 'Microsoft.AspNetCore.Identity.UserManager`1[Microsoft.AspNetCore.Identity.IdentityUser]' has been registered.**

Please help me to solve the problem

**Comment by JonPSmith** [comment]
Hi @nasir-adappt,

The AuthP library implements its own Roles system - see [this article](https://www.thereformedprogrammer.net/finally-a-library-that-improves-role-authorization-in-asp-net-core/) for how this works. That means you don't use the Individual Account security Roles.

You get the error because your individual user accounts database doesn't match the database that the `'Microsoft.AspNetCore.Identity.UserManager` expects.

PS. There are some useful documentation on Permissions and AuthP Roles in the `Concepts` section on the right had side of the [AuthP docs](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki).

**Reply by raulidavid** [reply]
Hello @JonPSmith , I also have the same problem, what do you mean by the fact that the database does not match???

services.RegisterAuthPermissions<Example2Permissions>(options =>
        {
            //This tells AuthP that you don't have multiple instances of your app running,
            //so it can run the startup services without a global lock
            options.UseLocksToUpdateGlobalResources = false;

            //This sets up the JWT Token. The config is suitable for using the Refresh Token with your JWT Token
            options.ConfigureAuthPJwtToken = new AuthPJwtConfiguration
            {
                Issuer = jwtData.Issuer,
                Audience = jwtData.Audience,
                SigningKey = jwtData.SigningKey,
                TokenExpires = new TimeSpan(0, 5, 0), //Quick Token expiration because we use a refresh token
                RefreshTokenExpires = new TimeSpan(1, 0, 0, 0) //Refresh token is valid for one day
            };
        })
        .UsingEfCoreSqlServer(configuration.GetConnectionString("SqlServerConnection")) //NOTE: This uses the same database as the individual accounts DB
        .IndividualAccountsAuthentication<ApplicationUser>()
        .AddSuperUserToIndividualAccounts<ApplicationUser>()
        .RegisterFindUserInfoService<CustomUserLookup>()
        .AddRolesPermissionsIfEmpty(AppAuthSetupData.RolesDefinition)
        .AddAuthUsersIfEmpty(AppAuthSetupData.UsersRolesDefinition)
        .SetupAspNetCoreAndDatabase(options =>
        {
            options.RegisterServiceToRunInJob<StartupServiceMigrateAnyDbContext<IdentityLocalDbContext>>();
            options.RegisterServiceToRunInJob<StartupServicesIndividualAccountsAddDemoUsers>();
        });

        services.AddScoped<IAddNewUserManager, IndividualUserAddUserManager<ApplicationUser>>();


System.InvalidOperationException: 'No service for type 'Microsoft.AspNetCore.Identity.UserManager`1[Microsoft.AspNetCore.Identity.IdentityUser]' has been registered.'

**Reply by JonPSmith** [reply]
hi @raulidavid,

I assume you get this exception on startup so the `IndividualUserAddUserManager<ApplicationUser>` doesn't match to what it expected. I can't be exactly sure why, but maybe your `ApplicationUser` doesn't match the `IdentityUser` class. 

Try comment out the `services.AddScoped<IAddNewUserManager, IndividualUserAddUserManager<ApplicationUser>>();` and check it works without that service. 

PS. The Example3.MvcWebApp.IndividualAccounts's Program uses the IndividualUserAddUserManager and works, so have a look if anything else is different in that project.

**Reply by raulidavid** [reply]
Thanks @JonPSmith I was able to solve it
